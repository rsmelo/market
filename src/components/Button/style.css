.button {
  font-family: var(--font-family);
  cursor: pointer;
  font-size: var(--font-size-default);
  text-align: center;
  border-radius: 4px;
  outline: 0;
  padding: 12px 14px;
  transition: 0.2s ease-out;
  border-style: solid;
  border-width: 1px;
  text-transform: uppercase;
}

.flat-green,
.flat-grey {
  color: var(--color-white);
}

.flat-green {
  background: var(--color-light-green);
  border-color: var(--color-light-green);
}

.flat-green:hover {
  background: var(--color-dark-green);
  border-color: var(--color-dark-green);
}

.flat-grey {
  background: var(--color-ligther-grey);
  border-color: var(--color-ligther-green);
}

.flat-grey:hover {
  background-color: var(--color-ligth-grey);
  border-color: var(--color-light-ligthgreen);
}

.outline {
  background: transparent;
}

.outline-green {
  color: var(--color-light-green);
  border-color: var(--color-light-green);
}

.outline-green:hover {
  color: var(--color-dark-green);
  border-color: var(--color-dark-green);
}

.outline-grey {
  color: var(--color-ligther-grey);
  border-color: var(--color-ligther-grey);
}

.outline-grey:hover {
  color: var(--color-ligth-grey);
  border-color: var(--color-ligth-grey);
}

.button:disabled:hover,
.button:disabled {
  background: var(--color-ligther-grey);
  color: var(--color-ligth-grey);
  border-color: var(--color-ligther-grey);
  cursor: not-allowed;
}
