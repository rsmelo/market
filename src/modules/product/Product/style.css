.product {
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
  width: 100%;
  padding: 1em 2em;
}

.image {
  max-width: 360px;
  max-height: 450px;
  border-radius: 3px;
}

.info {
  width: 550px;
  padding: 0 2em;
  margin-left: 2em;
  border-radius: 3px;
  background-color: var(--color-white);
}

.name {
  position: relative;
  overflow: hidden;
  white-space: nowrap;
  display: block;
  font-size: 1.6em;
  line-height: 1.4em;
  color: var(--color-dark-grey);
  text-transform: uppercase;
  margin: 1em 0 0;
}

.category {
  color: var(--color-ligther-grey);
}

.description {
  color: var(--color-ligth-grey);
  text-align: justify;
}

.price {
  font-size: 2.3em;
}

.priceContainer {
  display: flex;
  justify-content: space-between;
  flex-wrap: wrap;
}

.priceContainer,
.sellerContainer {
  padding-top: 1em;
}

.sellerContainer {
  color: var(--color-ligth-grey);
}

.buttonContainer {
  display: flex;
  flex-flow: column wrap;
}

.buttonContainer > a {
  margin-top: 0.5em;
}

.seller {
  font-weight: bold;
}
