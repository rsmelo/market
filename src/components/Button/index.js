import React from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'

import style from './style.css'

const Button = ({
  onClick,
  type,
  children,
  color,
  appearance,
}) => {
  const styles = classNames(
    style.button,
    style[appearance],
    style[`${appearance}-${color}`]
  )

  return (
    <button
      className={styles}
      onClick={onClick}
      type={type}
    >
      {children}
    </button>
  )
}


Button.defaultProps = {
  onClick: null,
  appearance: 'flat',
  color: 'green',
  type: 'button',
}

Button.propTypes = {
  type: PropTypes.oneOf(['submit', 'reset', 'button']),
  onClick: PropTypes.func,
  appearance: PropTypes.oneOf([
    'flat', 'outline',
  ]),
  color: PropTypes.oneOf([
    'green', 'grey',
  ]),
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.string,
    PropTypes.node,
  ]).isRequired,
}

export default Button
