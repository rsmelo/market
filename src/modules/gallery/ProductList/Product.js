import React from 'react'
import PropTypes from 'prop-types'
import { Link } from 'react-router-dom'

import style from './Product.style.css'

const Product = ({
  id,
  name,
  price,
  image,
}) => (
  <li className={style.product}>
    <Link to={`/product/${id}`}>
      <img
        className={style.image}
        src={image}
        alt={name}
      />
    </Link>
    <div className={style.information}>
      <Link to={`/product/${id}`}>
        <h2 className={style.name}>{name}</h2>
        <span className={style.price}>R$ {price}</span>
      </Link>
    </div>
  </li>
)

Product.propTypes = {
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  price: PropTypes.number.isRequired,
  image: PropTypes.string.isRequired,
}

export default Product
