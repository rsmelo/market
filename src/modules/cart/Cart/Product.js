import React from 'react'
import PropTypes from 'prop-types'

import Button from '../../../components/Button'
import style from './Product.style.css'

const Product = ({
  id,
  image,
  name,
  price,
  seller,
  onRemove,
}) => (
  <tr>
    <td>
      <img
        src={image}
        alt={name}
        className={style.image}
      />
    </td>
    <td className={style.productCell}>{name}</td>
    <td className={style.productCell}>R$ {price}</td>
    <td className={style.productCell}>
      <Button onClick={() => onRemove(id, seller.id)}>
        Remover do carrinho
      </Button>
    </td>
  </tr>
)

Product.propTypes = {
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  price: PropTypes.number.isRequired,
  image: PropTypes.string.isRequired,
  seller: PropTypes.shape({
    id: PropTypes.string.isRequired,
  }).isRequired,
  onRemove: PropTypes.func.isRequired,
}
export default Product
