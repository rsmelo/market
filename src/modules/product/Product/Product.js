import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'

import Loader from '../../../components/Loader'
import Button from '../../../components/Button'
import ButtonLink from '../../../components/ButtonLink'
import style from './style.css'

class Product extends PureComponent {
  constructor (props) {
    super(props)
    this.renderButtons = this.renderButtons.bind(this)
    this.renderProduct = this.renderProduct.bind(this)
  }

  componentDidMount () {
    const {
      product,
      match,
      getProduct,
    } = this.props
    if (!product) {
      getProduct(match.params.id)
    }
  }

  renderProduct () {
    const { product } = this.props
    return product && ([
      <div key={0} >
        <ButtonLink to="/" appearance="outline">
          Ver outros produtos
        </ButtonLink>
      </div>,
      <div className={style.imageContainer} key={1}>
        <img
          className={style.image}
          src={product.image}
          alt={product.name}
        />
      </div>,
      <div className={style.info} key={2}>
        <h2 className={style.name}>{product.name}</h2>
        <span className={style.category}>{product.category}</span>
        <p className={style.description}>{product.description}</p>
        <div className={style.priceContainer}>
          <span className={style.price}>R$ {product.price}</span>
          <div className={style.buttonContainer}>
            {this.renderButtons()}
            <ButtonLink to="/cart" appearance="outline">
              ir para o carrinho
            </ButtonLink>
          </div>
        </div>
        <div className={style.sellerContainer} key={3}>
          vendido por <span className={style.seller}>{product.seller.name}</span>
        </div>
      </div>,
    ])
  }

  renderButtons () {
    const {
      product,
      addedToCart,
      removeProduct,
      addProduct,
    } = this.props

    if (addedToCart) {
      return (
        <Button
          color="grey"
          appearance="outline"
          onClick={() => removeProduct(product.id, product.seller.id)}
        >
          Remover do carrinho
        </Button>
      )
    }
    return (
      <Button
        onClick={() => addProduct(product)}
      >
        Adicionar ao carrinho
      </Button>)
  }

  render () {
    const { isFetching } = this.props
    return (
      <div className={style.product}>
        {isFetching && <Loader />}
        {this.renderProduct()}
      </div>
    )
  }
}

Product.defaultProps = {
  product: null,
}

Product.propTypes = {
  product: PropTypes.shape({
    id: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    price: PropTypes.number.isRequired,
    image: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    category: PropTypes.string.isRequired,
    seller: PropTypes.shape({
      id: PropTypes.string.isRequired,
      name: PropTypes.string.isRequired,
    }),
  }),
  addedToCart: PropTypes.bool.isRequired,
  match: PropTypes.shape({
    params: PropTypes.shape({
      id: PropTypes.string,
    }),
  }).isRequired,
  isFetching: PropTypes.bool.isRequired,
  getProduct: PropTypes.func.isRequired,
  addProduct: PropTypes.func.isRequired,
  removeProduct: PropTypes.func.isRequired,
}
export default Product
