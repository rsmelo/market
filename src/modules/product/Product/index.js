import { connect } from 'react-redux'
import { withRouter } from 'react-router-dom'

import Product from './Product'

import {
  getGalleryProductById,
  getGalleryIsFetching,
  getExistProductInCart,
} from '../../combinedReducers'
import { getProduct } from '../actions'
import { addProduct, removeProduct } from '../../cart/actions'

const mapStateToProps = (state, ownProps) => {
  const product = getGalleryProductById(state, ownProps.match.params.id)
  return {
    product,
    isFetching: getGalleryIsFetching(state),
    addedToCart: product ?
      getExistProductInCart(state, product.id, product.seller.id) :
      false,
  }
}

const mapDispatchToProps = {
  getProduct,
  addProduct,
  removeProduct,
}

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(Product))
