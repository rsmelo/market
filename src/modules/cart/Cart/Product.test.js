import React from 'react'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'

import Product from './Product'
import Button from '../../../components/Button'

jest.mock('react-router-dom', () => ({ Link: 'Link' }))

describe('Product', () => {
  it('should render correctly', () => {
    const product = {
      id: 'dd5ca2b6-b8ae-4627-95f3-7d49613a0df5',
      name: 'Abominável',
      price: 80,
      image: 'https://images-na.ssl-images-amazon.com/images/I/A1eI8aYKX%2BL._SL1500_.jpg',
      description: 'description',
      category: 'action figures',
      seller: {
        id: 're_cjabsxlgq01t5oq6f3ix79dsn',
        name: 'João',
      },
    }
    const removeProduct = jest.fn()
    const tree = renderer
      .create(<Product {...product} onRemove={removeProduct} />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should have a Product to each item in products property', () => {
    const product = {
      id: 'dd5ca2b6-b8ae-4627-95f3-7d49613a0df5',
      name: 'Abominável',
      price: 80,
      image: 'https://images-na.ssl-images-amazon.com/images/I/A1eI8aYKX%2BL._SL1500_.jpg',
      description: 'description',
      category: 'action figures',
      seller: {
        id: 're_cjabsxlgq01t5oq6f3ix79dsn',
        name: 'João',
      },
    }
    const removeProduct = jest.fn()
    const component = shallow(<Product {...product} onRemove={removeProduct} />)
    component.find(Button).simulate('click')
    expect(removeProduct).toHaveBeenCalledWith(product.id, product.seller.id)
  })
})
