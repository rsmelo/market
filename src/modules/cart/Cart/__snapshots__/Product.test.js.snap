// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product should render correctly 1`] = `
<tr>
  <td
    className=""
  >
    <img
      alt="Abominável"
      className={undefined}
      src="https://images-na.ssl-images-amazon.com/images/I/A1eI8aYKX%2BL._SL1500_.jpg"
    />
  </td>
  <td
    className=""
  >
    Abominável
  </td>
  <td
    className=""
  >
    R$ 
    80
  </td>
  <td
    className=""
  >
    <button
      className=""
      onClick={[Function]}
      type="button"
    >
      Remover do carrinho
    </button>
  </td>
</tr>
`;
